### DISCUSSION TITLE
How to check types for the entire project via CLI?

### DISCUSSION THREAD

**Discussion by jaydenseric** [denoland/deno#12113] [Q&A]
There is [an issue about introducing a `deno check` CLI command](https://github.com/denoland/deno/issues/11786), which presumably would allow TypeScript types to be checked for the entire project via CLI (for CI), but that could be weeks/months away from a resolution so how do we do it today?

**Comment by kitsonk** [comment]
`deno cache`

Deno currently does not have a concept of a "project".  If you `deno cache <somefile>`, Deno will analyse its dependencies and type check them.  If you have multiple entry points, you would do each.

Also, `deno test` is another option, where you have at least a file that Deno recognises at a test module that imports all the modules in your project.  Then you don't have to specify anything other than `deno test`.

**Reply by jaydenseric** [reply]
It sounds like it's impossible out of the box with the Deno CLI then 🙃

The suggestions to create an entrypoint module that imports every other JS and TS module in the project is not a viable solution. Firstly, it would be huge and difficult to write and maintain. Secondly, the point of this is to have a CLI way to check for human errors. What if the human error is that files are missing from this manually curated entry module? CI would just not notice anything is wrong.

Secondly, what happens if there are type errors in several of the modules imported in the entry module? Would it crash at the first error, or would it continue to check all of the modules and print a report containing every error, and not just the first error?

To survive in the meantime, is the best way forward to make a custom CLI that globs up all the `.js`/`.ts`, etc. files and attempts to run them with the `deno cache` command via new processes, and somehow forward any stdout/stderr to the current console?

This sounds fiddly, and will probably be a wasted effort once `deno check` or whatever the official command turns out to be is released.

A type check command is an MVP critical feature, how come it doesn't exist yet after several years? Does team Deno not check for type errors in CI? Or do you just assume 100% of the codebase gets pulled into tests and is checked that way? Surely there are scripts and things in Deno repos that are not a part of tests or the tested codebase that are not currently checked in CI.

**Reply by kitsonk** [reply]
Modules need to get imported somewhere to be used.

For applications, the entry point has conventionally been `main.ts`.  For libraries `mod.ts`.  For a library, a jumble of modules with no clear contracted set of public APIs to consume tends to be confusing to consumers and is generally considered an anti-pattern.  For more complex projects, there maybe multiple entry points, which is planned to be helped with the `deno.jsonc` configuration file.  But still Deno will likely not glob every file, but work specifically off a dependency graph of the code included in the project.

If your module doesn't get imported or tested, it doesn't get type checked.  Most CI pipelines run tests, ergo, code gets type checked.

**Reply by jaydenseric** [reply]
Index modules that export the entire API are absolutely an anti-pattern and it's disappointing to see Deno encourage this _(**Edit 24 July 2022:** See this article [“Optimal JavaScript module design”](https://jaydenseric.com/blog/optimal-javascript-module-design))_. The public API should be modules that are deep-imported, so that the user only downloads, parses and runs the necessary code. People in Deno/Node.js world got used to the convenience of having a single index to import everything from, but this pattern has basically destroyed the quality of the web in browsers. People bundling/loading in megabytes of JS just to use one or two functions. Many websites load all of `lodash` just to use one or two utilities, or they assume/pray their bundler "tree-shakes" away all of the dead code their app pulls in with laptop fans whirring every build. Here's an idea; just the import code you intend to use. See this thread on exactly this topic:

https://twitter.com/jaydenseric/status/1410438154541146113

Deno projects particularly benefit from the deep-import approach because if you import from an index module that exports a bunch of undesired things imported from other modules, you end up waiting for them all to be fetched and cached, slowing down script cache/start times and bloating the size of your cache on disk with code you're not actually using.

> If your module doesn't get imported or tested, it doesn't get type checked. Most CI pipelines run tests, ergo, code gets type checked.

This is a leaky approach. Here is an exact example of a situation where a project has a complex JS module that would never get imported or tested in a CI run:

https://github.com/jaydenseric/device-agnostic-ui-website/blob/master/scripts/screenshots.mjs

It's a script that uses `puppeteer` to generate UI library component screenshot assets in the Next.js `public/static/screenshots` directory. I only manually run it locally after modifying a particular component in a way that changes it's appearance. If I migrated that project to Deno, and attempted type safety for all the JS/TS code, CI would not notice type errors in the screenshot script:

https://github.com/jaydenseric/device-agnostic-ui-website/blob/f1ef8247054bd8847269b3feec7079e3f6bf7014/.github/workflows/ci.yml#L18

CI is not the only use case for a type-checking CLI command. At this very moment, I'm trying to find all the problems in my work-in-progress project so I can figure out the next batch of type issues to work through and fix. While working on a project like I am right now, often 50% of the modules you have written don't have any tests yet (but will eventually). I want to be able to type check these modules *before* focusing on testing them.

I wrote a shell script to attempt to type check all project modules:

```
#!/bin/zsh

for i in **/*.{js,ts}
do
  deno cache --import-map importMapDev.json "$i"
done
```

I'm not sure yet because the output is so noisy, but there appears to be a problem with it; type errors are not deduped. If a module with a type error is imported 20 times by other modules, then that same type error will be spammed in output 20+ times.

**Reply by nayeemrmn** [reply]
Also see `deno test --no-run` for blanket source detection, it exists so users can `deno cache` all relevant modules in a project, with the requirement that any such module is at least referenced by some `foo_test.ts` file. Encourages having tests I guess. If you need to work outside of that constraint then there's nothing wrong with shell-level globs / external tools as you have attempted. To dedup type errors ensure that the glob is expanded positionally in a single `deno cache **/*.{js,ts}` command.

`deno check` is irrelevant to this issue, it'll more or less just be what `deno cache` is now. I suppose you should propose that `deno cache` without arguments would scan for all source files similarly to `deno lint` (that change would be inherited by `deno check`).

**Reply by jaydenseric** [reply]
@nayeemrmn you legend! Your suggestion unblocked me:

```sh
deno cache --import-map importMapDev.json **/*.{js,ts}
```

People will generally want to check for all issues in the project at once, it would be really neat if there was a single Deno command to check all project files for fmt, lint, type, and test errors grouping all problems in output under each file. Here is my GitHub Actions CI config now:

```yml
name: CI
on: [push, pull_request]
jobs:
  test:
    name: Test with ${{ matrix.os }}
    runs-on: ${{ matrix.os }}
    strategy:
      matrix:
        os: [ubuntu-latest, macos-latest]
    steps:
      - name: Checkout repo
        uses: actions/checkout@v2
      - name: Setup Deno
        uses: denoland/setup-deno@v1
        with:
          deno-version: '^1.0.0'
      - name: Format check
        run: deno fmt --check
      - name: Lint
        run: deno lint
      - name: Type check
        run: deno cache --import-map importMapDev.json **/*.{js,ts}
      - name: Test
        run: deno test --import-map=importMapDev.json --allow-net
```

I don't use `.jsx`/`.tsx` files anymore, but anyone copying the config and glob should account for those file extensions too.

**Reply by jaydenseric** [reply]
I suppose the order matters too; do the `deno cache` based type check before other checks like `deno test`, that way caching isn't done twice and benefits following commands.

**Reply by jaydenseric** [reply]
Ok, So I've been having nightmares all afternoon figuring out 1980's bullshit technology to get the glob to work in all the different shells, and it looks like I'm getting close to a solution with this monstrosity:

```sh
find . \( -name  '*.js' -o  -iname '*.jsx' -o  -iname '*.ts' -o  -iname '*.tsx' \) -print0 | xargs -0 deno cache --import-map importMapDev.json
```

Some of the challenges are:

- You don't want the glob to resolve to the string of the glob itself if no files are found.
- If files for one of the possible file extensions isn't found (but others are) you don't want an error.

If ZSH were available everywhere, you could do things like this:

```sh
#!/bin/zsh
deno cache --import-map importMapDev.json -- **/*.(js|jsx|ts|tsx)
```

But it's not available in GitHub Actions and the `bash` shell in GitHub actions is only v3.2:

https://github.com/actions/virtual-environments/blob/main/images/macos/macos-11-Readme.md#language-and-runtime

So you can't use `**` in globs, or any of the optional glob settings about what happens if no results are found.

While a full-blown Deno script to do the same thing as my final monstrosity shell script using std lib functions for globs, etc. would be more code and slower, it would be way more intuitive to understand. I've spent the whole afternoon on this and I'm still not happy with the result, and totally distracted from writing actual project JS code. This is not the developer experience Deno aspires to :(

**Reply by jaydenseric** [reply]
I've just discovered another problem with using `deno cache` to type check; it is unable to discover type errors in `.js` files that have the `// @ts-check` comment, even though the Deno LSP shows type errors in these files in your editor.

**Reply by jaydenseric** [reply]
I raised a bug issue for the above: https://github.com/denoland/deno/issues/12143 .

**Comment by jaydenseric** [comment]
Here is what I came up with:

https://github.com/jaydenseric/ruck/commit/ef8275bc0c739ef2435d9430aa13f82d4f8c7eef

```js
#!/usr/bin/env -S deno run --allow-read --allow-write --allow-run

// @ts-check
// Type checks every JavaScript module in the project.

import { expandGlob } from "std/fs/mod.ts";

const typeCheckIndexModuleName = "type-check-index.mjs";

console.log("Creating a temporary index module to check…");

let rawModule = "";

for await (const { path } of expandGlob("**\/*.mjs")) {
  rawModule += `import "${path}";\n`;
}

await Deno.writeTextFile(typeCheckIndexModuleName, rawModule);

console.log("Checking the temporary index module…");

const process = Deno.run({
  cmd: ["deno", "check", typeCheckIndexModuleName],
  stdout: "piped",
  stderr: "piped",
});

const { code } = await process.status();
const rawOutput = await process.output();
const rawError = await process.stderrOutput();

if (code === 0) await Deno.stdout.write(rawOutput);
else await Deno.stderr.write(rawError);

console.log("Deleting the temporary index module…");

await Deno.remove(typeCheckIndexModuleName);

Deno.exit(code);
```

It can be run in the project directory by running this:

```sh
./scripts/type-check.mjs
```

**Comment by jollytoad** [comment]
I think the behaviour of least surprise would be for `deno check` and `deno cache` to operate on all files that `deno fmt` and `deno lint` do when given no specific files.

I find myself running the following quite often:

```
deno fmt && deno lint && deno check **/*.ts
```

And having to explain this to other devs, and hope their shell supports globs.

It's frustrating that can't be added as a task, as tasks don't yet support globs.

It would be far more intuitive to just be able to do:

```
deno fmt && deno lint && deno check
```

Which could also easily be added as a task too.

**Reply by jollytoad** [reply]
The default operation of check could also be configured in the same way as fmt/lint to allow a set of file includes/excludes.

Consistency is the key to a great user experience, and this current inconsistency doesn't benefit anyone.

**Reply by deer** [reply]
`deno check **/*.ts` is very useful -- thanks for posting this!

Since it's now almost seven months in the future, just thought I'd share the good news here for anyone else finding this. Deno now _does_ support globs from tasks: https://deno.com/blog/v1.34#glob-support-in-denojson-and-cli-flags

So you could add something like this to your tasks:
`"safe": "deno fmt --check && deno lint && deno check **/*.ts"`

**Comment by lytedev** [comment]
Is this something that a PR would be welcome for? I also think `deno check` should work similarly to `deno fmt`, `deno lint`, and `deno test`. Especially since globbing does not yet work in deno tasks and I want to do something like this:

```jsonc
{
	"tasks": {
		"pre-commit": "deno lint && deno check --all *.ts && deno test"
		"setup-pre-commit": "echo '#!/bin/sh\ndeno task pre-commit' > .git/hooks/pre-commit && chmod +x .git/hooks/pre-commit"
	}
}
```

**Comment by hrumhurum** [comment]
`deno check **/*.ts` does not work on Windows:

`error: The filename, directory name, or volume label syntax is incorrect. (os error 123)`

It would be useful if `deno check` could do this by default. Alternatively, `deno check` would benefit from a built-in glob support, as not all shell interpreters provide it.

**Reply by bartlomieju** [reply]
@hrumhurum this should work if you wrap `**/*.ts` in quotes like so: `deno check "**/*.ts"` - then Deno will expand the glob itself.

**Reply by hrumhurum** [reply]
`deno check "**/*.ts"` does not work and produces the same error message on Windows.

**Reply by greentore** [reply]
It *does* work if you define it as a task however.

```json
  "tasks": {
    "check": "deno check **/*.ts"
  }
  ```

**Comment by silverwind** [comment]
On GitHub actions, I needed to use `shopt -s globstar; deno check src/**/*.ts` because the bash shell there does not have recursive globs enabled by default.

I think deno should support `deno check` (current directory) and `deno check <dir>` and just discover all `.ts` and `.tsx` files itself.
